solr/core/src/java/org/apache/solr/cloud/api/collections/CreateCollectionCmd.java

```diff
@@ -590,12 +589,22 @@ public static List<String> populateShardNames(ZkNodeProps message, String router
     return shardNames;
   }
 
-  private static ReplicaCount getNumReplicas(ZkNodeProps message) {
-    int numTlogReplicas = message.getInt(TLOG_REPLICAS, 0);
-    int numNrtReplicas =
-        message.getInt(
-            NRT_REPLICAS, message.getInt(REPLICATION_FACTOR, numTlogReplicas > 0 ? 0 : 1));
-    return new ReplicaCount(numNrtReplicas, numTlogReplicas, message.getInt(PULL_REPLICAS, 0));
+  private ReplicaCount getNumReplicas(ZkNodeProps message) {
+    ReplicaCount numReplicas = ReplicaCount.fromMessage(message);
+    int numLeaderEligibleReplicas = numReplicas.count(leaderEligibleReplicaTypes);
+    if (numLeaderEligibleReplicas == 0 && !numReplicas.contains(Replica.Type.defaultType())) {
+      // Ensure that there is at least one replica that can become leader if the user did
+      // not force a replica count.
+      numReplicas.put(Replica.Type.defaultType(), 1);
+    } else if (numLeaderEligibleReplicas == 0) {
+      // This can still fail if the user manually forced "0" replica counts.
```

Comment:
This slightly changes the existing behaviour. As noted in the comment, if a user explicitly passes `nrtReplicas=0` to the create collection command, it would fail. With your suggestion, we would silently force `nrtReplicas=1`. I don't mind changing the behaviour, but I tried very hard not to.

Reply:
And yes `ReplicaCount#validate` was useless, that's why I removed it. ;)

Reply:
The current behaviour is:
- if I pass no parameters, I will have 1 NRT replica
-  if I pass `pullReplicas=1`, I will have 1 NRT and 1 PULL replica
- if I pass `nrtReplicas=0`, I will have an error

What you are suggesting is to change the third situation, and instead to silently transform the user request into `nrtReplicas=1`. The error is triggered if the default replica type is explicitly set to 0, not if all replica types are set to 0.

Reply:
I don't think this PR should change behavior like that.

FYI it's valid to create a collection with no replicas!  The expectation is that the client will explicitly add them afterwards, likely at specific nodes of its choosing.

Reply:
I don't think it is valid to create a collection with 0 replicas. I went back to the commit before I introduced `ReplicaCount`, and I see [the following](https://github.com/apache/solr/blob/8b9954317690428b13e1c6c4ed1dbdb743ab389b/solr/core/src/java/org/apache/solr/cloud/api/collections/CreateCollectionCmd.java#L580):

```java
int numTlogReplicas = message.getInt(TLOG_REPLICAS, 0);
int numNrtReplicas =
    message.getInt(
        NRT_REPLICAS, message.getInt(REPLICATION_FACTOR, numTlogReplicas > 0 ? 0 : 1));

if (numNrtReplicas + numTlogReplicas <= 0) {
  throw new SolrException(
      ErrorCode.BAD_REQUEST, NRT_REPLICAS + " + " + TLOG_REPLICAS + " must be greater than 0");
}
```

Reply:
See the docs: https://solr.apache.org/guide/solr/latest/deployment-guide/collection-management.html `createReplicas` boolean for v2 api (v1 has an equivalent means as well).

Reply:
Well, then it works because if you pass `createReplicas=false` alone,  this will default to`nrtReplicas=1`. I am pretty sure that if you explicitly pass `createReplicas=false&nrtReplicas=0` the call is rejected.

Reply:
This succeeds: ```
curl -X POST http://localhost:8983/api/collections -H 'Content-Type: application/json' -d '
  {
    "name": "techproducts_v2",
    "config": "_default",
    "numShards": 1,
    "createReplicas": false
  }'
```
I run Solr with `docker run --rm -p 8983:8983 solr:9 -c`

Reply:
It may be a bug (what is the purpose of validating the number of replicas if not creating any replicas?), but the goal of this PR is not to change this behaviour.

Reply:
It's a small change in behavior for an edge-case.  Wouldn't break any existing user.